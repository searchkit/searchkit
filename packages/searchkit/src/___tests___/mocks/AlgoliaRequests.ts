export const SimpleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['type:movie']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  }
]

export const nonDynamicFacetRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facets: ['type', 'actors', 'imdbrating'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const nonDynamicFacetRequestOneFilter = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['actors:Alice']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const DisjunctiveExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['type:movie']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'type',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  }
]

export const NestedQueryRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['user.first:Alice']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'user.first',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const TwoFiltersNestedQueryRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['user.first:Alice', 'user.first:John']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'user.first',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const NumericFilterNestedQueryRequest = [
  {
    indexName: 'my-index-000001',
    params: {
      facetFilters: [],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      numericFilters: ['user.price>=90'],
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const FacetQueryExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetName: 'type',
      facetQuery: 'm',
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxFacetHits: 10,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const NumericFiltersExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facets: ['type', 'actors', 'imdbrating'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      numericFilters: ['imdbrating>=90'],
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      clickAnalytics: false,
      facets: 'imdbrating',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 0,
      maxValuesPerFacet: 10,
      page: 0,
      query: ''
    }
  }
]

export const NestedFacetQueryExampleRequest = [
  {
    indexName: 'my-index-000001',
    params: {
      facetName: 'user.first',
      facetQuery: 'har',
      facets: ['user.first', 'user.price'],
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      maxFacetHits: 10,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]
