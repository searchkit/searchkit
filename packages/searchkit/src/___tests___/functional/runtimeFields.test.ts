import Client from '../../'
import { DisjunctiveExampleRequest, SimpleRequest } from '../mocks/AlgoliaRequests'
import type { AlgoliaMultipleQueriesQuery } from '@searchkit/api'

import nock from 'nock'
import {
  HitsResponseWithFacetFilter,
  HitsResponseWithNoFacets
} from '../mocks/ElasticsearchResponses'

describe('Integration tests for runtime mapping fields', () => {
  it('call with one filter and query applied', async () => {
    const client = new Client({
      connection: {
        host: 'http://localhost:9200',
        apiKey: 'a2Rha1VJTUJMcGU4ajA3Tm9fZ0Y6MjAzX2pLbURTXy1hNm9SUGZGRlhJdw=='
      },
      search_settings: {
        runtime_mappings: {
          rating: {
            type: 'keyword',
            script: {
              source: "emit(doc['rated'].size()>0 ? doc['rated'].value : '')"
            }
          }
        },
        search_attributes: ['rating'],
        result_attributes: ['rating'],
        filter_attributes: [{ attribute: 'type', field: 'type', type: 'string' }]
      }
    })

    nock('http://localhost:9200')
      .post('/_msearch', (requestBody: any) => {
        expect(requestBody).toMatchSnapshot('ES Request')
        return true
      })
      .reply(200, HitsResponseWithNoFacets)

    const response = await client.handleInstantSearchRequests(
      SimpleRequest as AlgoliaMultipleQueriesQuery[]
    )

    expect(response).toMatchSnapshot()
  })
})
