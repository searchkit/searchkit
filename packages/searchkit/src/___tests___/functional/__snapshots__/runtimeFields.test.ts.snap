// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration tests for runtime mapping fields call with one filter and query applied 1`] = `
{
  "results": [
    {
      "appliedRules": [],
      "exhaustive": {
        "facetsCount": true,
        "nbHits": true,
        "typo": true,
      },
      "exhaustiveFacetsCount": true,
      "exhaustiveNbHits": true,
      "exhaustiveTypo": true,
      "hits": [
        {
          "_index": "imdb_movies",
          "_score": 7.637228,
          "actors": [
            "Tim Robbins",
            "Morgan Freeman",
            "Bob Gunton",
            "William Sadler",
          ],
          "objectID": "tt0111161",
          "rated": [
            "R",
          ],
          "title": "The Shawshank Redemption",
        },
      ],
      "hitsPerPage": 20,
      "index": "imdb_movies",
      "nbHits": 1,
      "nbPages": 1,
      "page": 0,
      "params": "facetFilters=type%3Amovie&facets=*&highlightPostTag=%3C%2Fem%3E&highlightPreTag=%3Cem%3E&maxValuesPerFacet=10&page=0&query=shawshank&tagFilters=",
      "processingTimeMS": 2,
      "query": "shawshank",
      "renderingContent": {
        "facetOrdering": {
          "facets": {
            "order": [],
          },
          "values": undefined,
        },
      },
    },
  ],
}
`;

exports[`Integration tests for runtime mapping fields call with one filter and query applied: ES Request 1`] = `
"{"index":"imdb_movies"}
{"aggs":{},"query":{"bool":{"filter":[{"bool":{"should":[{"term":{"type":"movie"}}]}}],"must":{"bool":{"should":[{"bool":{"should":[{"multi_match":{"query":"shawshank","fields":["rating"],"fuzziness":"AUTO:4,8"}},{"multi_match":{"query":"shawshank","fields":["rating"],"type":"bool_prefix"}}]}},{"multi_match":{"query":"shawshank","type":"phrase","fields":["rating"]}}]}}}},"size":20,"from":0,"_source":{"includes":["rating"]},"fields":["rating"],"runtime_mappings":{"rating":{"type":"keyword","script":{"source":"emit(doc['rated'].size()>0 ? doc['rated'].value : '')"}}}}
"
`;
